### ISSUE TITLE
Nullability doesn't recognize assignment to new objects after null check

### ISSUE THREAD

**Issue by praeclarum** [dotnet/roslyn#31802]
In the following code:

```csharp
PolygonTreeNodeList? backnodes = ...;
if (backnodes == null) backnodes = new PolygonTreeNodeList();
backnodes.Add(this); // CS8602
```

The error is raised even though the data flow seems pretty straightforward. It would be nice if no warning was emitted here.

**Version Used**: .NET Core 3.0.100-preview-009812

**Expected Behavior**: Compiles OK.

**Actual Behavior**: Error CS8602

**Comment by jcouv** [comment]
I wasn't able to repro. Here's what I tried:
```C#
#nullable enable
class C
{
    void M(object? o)
    {
        if (o == null) o = new object();
        o.ToString(); // no warning, as expected
    }
}
```
